### PR TITLE
f/membership-config

### PR DESCRIPTION
**Pull request by michael-scheurer**

### CODE REVIEW

File: libs/api/prisma/schema.prisma [hunk 1]
```diff
@@ -982,3 +982,59 @@ model TaggedEvents {
   @@id([eventId, tagId])
   @@map("events.tagged-events")
 }
+
+model SubscriptionTimeline {
+	id                  String @id @default(cuid())
```
Comment (Itrulia) [inline libs/api/prisma/schema.prisma:987]: We are in the process of switching to GUID for everything, could we move that to GUID already? the Event model above should have an example.

**Comment by michael-scheurer** [comment]
First UI feedback

- [ ] cannot create payment method
- [ ] cannot create a new subscription flow
- [ ] cannot delete custom event within a subscription flow
- [ ] within a subscription flow of a member plan: adjust the title, so that the name of the memberplan is visible (not: standard member plan)
image

- [ ] Add amount of affected subscription per subscription flow (see miro)
image

- [ ] In the mailtemplate synchronisation list, please add icon to "not in use" hint.
- [ ] Fix the scroll bars here 
image

- [ ] cannot move the custom event here 
image

- [ ] Fix margin/padding 
image

- [ ] would not show the vertical lines unless an event is placed above or below or dragging is happening. 
image

- [ ] instead of "add new" I would add an icon (like the delete). And give a bit more space to the left, to the flow. It sticks too much for now.
image

- [ ] Every event should have it's color and an icon.
image

- [ ] Every event should have an info icon, where a short explanation popover is implemented.
image

- [ ] Can we make the red parts sticky when scrolling?
image

- [ ] The defined events and the filter part are looking too similar. The filter should look more like a filter. For the events I would try to re-use the framed events of the individual flow. That would also uniform the use of color and icon. Additionally, we should think about, moving the defined events, not time related, moving in one column of the table and place them as a grid. That would save some horizontal scrolling.
image

- [ ] Cannot read the mail template within an event. Should find a solution for that. Eventually show the template only as text but with an edit button and only display the drop-down if in edit mode. Could be a possible way to go.
image

File: libs/membership/api/src/lib/mail/mail.controller.ts [hunk 2]
```diff
@@ -0,0 +1,93 @@
...
+  UserFlow
+}
+
+export type MailControllerConfig = {
+  daysAwayFromEnding?: number | null
+  externalMailTemplateId: string
+  recipient: User
+  isRetry?: boolean
+  periodicJobRunDate?: Date | null
+  optionalData: Record<string, any>
```
Comment (Itrulia) [inline libs/membership/api/src/lib/mail/mail.controller.ts:18]:
```suggestion
  optionalData: Record<string, unknown>
```

File: libs/membership/api/src/lib/periodic-job/periodic-job.executor.ts [hunk 3]
```diff
@@ -0,0 +1,40 @@
...
+    // Allow only chron that run once a day example: [number] [number] [number] [star] [star] [star]
+    {
+      disabled: !/[1-5]?[0-9] [1-5]?[0-9] [1-2]?[0-9] \* \* \*/.test(
+        process.env.PERIODIC_JOB_EXECUTION_SCHEDULE || `0 0 3 * * *`
+      )
+    }
+  )
+  async handleCron() {
+    console.log(
+      'Called when the current second is' +
```
Comment (Itrulia) [inline libs/membership/api/src/lib/periodic-job/periodic-job.executor.ts:24]: could use template strings

File: libs/membership/api/src/lib/subscription-event-dictionary/subscription-event-dictionary.ts [hunk 4]
```diff
@@ -0,0 +1,265 @@
...
+        daysAwayFromEnding: interval.daysAwayFromEnding,
+        externalMailTemplate: interval.externalMailTemplate
+      })
+
+      // Store earliest invoice creation date
+      if (interval.event === SubscriptionEvent.INVOICE_CREATION) {
+        if (
+          !this.earliestInvoiceCreationDate ||
+          interval.daysAwayFromEnding < this.earliestInvoiceCreationDate
+        )
```
Comment (Itrulia) [inline libs/membership/api/src/lib/subscription-event-dictionary/subscription-event-dictionary.ts:116]: Missing {}

File: libs/api/src/lib/mails/mailContext.ts [hunk 5]
```diff
@@ -72,53 +55,80 @@ export class MailContext implements MailContext {
...
   }
 
-  async sendMail({type, recipient, data}: SendEMailProps): Promise<void> {
-    // const type = SendMailType.LoginLink
-    const mailTemplate = this.mailTemplateMaps.find(template => template.type === type)
-    if (!mailTemplate) {
-      logger('mailContext').warn(`Template for type: ${type} not found`)
-      return
+  getProvider(): MailProvider {
+    if (!this.mailProvider) {
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply libs/api/src/lib/mails/mailContext.ts:101]: If the server would throw an error here if no mail provider is provided, then it probably should be non nullable imo. What do you think @tomaszdurka

File: libs/api/src/lib/mails/MailchimpMailProvider.ts [hunk 6]
```diff
@@ -111,78 +115,83 @@ export class MailchimpMailProvider extends BaseMailProvider {
...
+          merge_vars: [
+            {
+              rcpt: props.recipient,
+              vars: templateContent
+            }
+          ]
+        }
+      })
+      if (this.responseIsError(response)) {
+        return new MailProviderError(response.response?.data.message)
```
[the remark this answers is not in the document]
Comment (Itrulia) [reply libs/api/src/lib/mails/MailchimpMailProvider.ts:150]: Yeah, it's a shame that Typescript doesn't force callers to handle them.
The problem I have with returning an error is that if they are not handled then there's no error thrown.

I would rather have an error thrown than the possibility of it being forgotten. What do you think?

File: libs/membership/api/src/lib/mail-template/mail-template.resolver.ts [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+  async provider() {
+    const provider = await this.oldContextService.context.mailContext.getProvider()
+    return {name: provider.name}
+  }
+
+  @Permissions(CanSyncMailTemplates)
+  @Mutation(() => Boolean)
+  async syncTemplates() {
+    await this.syncService.synchronizeTemplates()
+    return true
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/membership/api/src/lib/mail-template/mail-template.resolver.ts:41]: Not necessarily a boolean, but the alternative is probably a `VoidScalar` from an external library: https://github.com/MichalLytek/type-graphql/issues/318#issuecomment-485070013

File: libs/membership/api/jest.config.ts [hunk 8]
```diff
@@ -13,5 +13,6 @@ export default {
   },
   moduleFileExtensions: ['ts', 'js', 'html'],
   coverageDirectory: '../../../coverage/libs/membership/api',
-  globalSetup: '<rootDir>/setup-database.js'
+  globalSetup: '<rootDir>/setup-database.js',
+  maxWorkers: 1
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/membership/api/jest.config.ts:17]: No, they don't depend on other tests, but running multiple tests in parallel breaks them, because they access the same database.

File: libs/api/src/lib/server.ts [hunk 9]
```diff
@@ -109,23 +116,11 @@ export class WepublishServer {
         res.status(500).end()
       }
     })
-
+    // Only as workaround until everything is migrated to NESTJS
+    global.oldContext = await contextFromRequest(null, this.opts)
     this.app = app
   }
 
-  async runJob(command: JobType, data: any): Promise<void> {
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/api/src/lib/server.ts:116]: The Job runner makes sure it only runs once per application: https://github.com/wepublish/wepublish/blob/f/membership-config/libs/membership/api/src/lib/periodic-job/periodic-job.controller.ts#L48
This should prevent it from running multiple times in a HA setup.

File: libs/membership/api/src/lib/periodic-job/periodic-job.controller.ts [hunk 10]
```diff
@@ -0,0 +1,462 @@
...
+   * @param startDate The beginning date (inclusive)
+   * @param endDate The ending date (exclusive)
+   * @returns An array of Date objects
+   */
+  private generateDateArray(startDate: Date, endDate: Date) {
+    const dateArray = []
+    const lastDate = this.getStartOfDay(endDate)
+    let inputDate = this.getStartOfDay(startDate)
+    while (inputDate < lastDate) {
+      inputDate = addDays(inputDate, 1)
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/membership/api/src/lib/periodic-job/periodic-job.controller.ts:437]: Yes, `date-fns` is not mutating the arguments (pedantic: dates are still reference types and `passed by reference`, but this is orthogonal to the mutation/non-mutation of a function)

File: libs/membership/api/jest.config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (Itrulia) [reply libs/membership/api/jest.config.ts:17]: If the tests clear the whole DB, they do rely on other tests not running, so there is a dependency. maxWorkers shouldn't be set to 1 (or use run in band). They should be able to run in parallel or independently.

I would rather not introduce this, as this will lower the quality of our tests  and in the future.

File: libs/membership/api/src/lib/subscription-event-dictionary/subscription-event-dictionary.ts [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+    intervals: (SubscriptionInterval & {mailTemplate: MailTemplate | null})[]
+  })[] = []
+
+  constructor(private readonly prismaService: PrismaClient) {}
+
+  /**
+   * Load all subscription flows. This is required before calling
+   * {@link getActionFromStore}.
+   */
+  public async initialize() {
```
Comment (tomaszdurka) [inline libs/membership/api/src/lib/subscription-event-dictionary/subscription-event-dictionary.ts:32]: Maybe just make it automatically run if it's not built?

File: libs/membership/api/src/lib/subscription-event-dictionary/subscription-event-dictionary.ts [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+  /**
+   * Try to get the MailTemplates for a specific filter of [memberPlan,
+   * paymentMethod, periodicity, autoRenewal]. You additionally need to pass
+   * either `daysAwayFromEnding` or `events` to filter further. If any
+   * filter leads to an empty result set, the templates of the default flow are
+   * returned.
+   * @param query The filter of the above properties.
+   * @returns An array of MailTemplates.
+   */
+  public getActionFromStore(query: LookupActionInput): Action[] {
```
Comment (tomaszdurka) [inline libs/membership/api/src/lib/subscription-event-dictionary/subscription-event-dictionary.ts:138]: Question: why method is called getActionFromStore if according to description it gets MailTemplates?
Also plural vs singular.

**Review by tomaszdurka — CHANGES_REQUESTED** [review]
Hey, this PR is huge, with a lot of changes from different perspectives. 
We need to address:
- allow to tests run in parrarel
- answers for our questions are really needed (!)
- would be great to separate mail changes and subscription flow changes

File: libs/membership/api/jest.config.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/membership/api/jest.config.ts:17]: @Itrulia @tomaszdurka 

> I would rather not introduce this, as this will lower the quality of our tests and in the future.

I understand, but we didn't introduce this. The problem already existed for example in [DashboardSubscriptionResolver](https://github.com/wepublish/wepublish/blob/f/membership-config/libs/membership/api/src/lib/dashboard/subscriptions/dashboard-subscription.resolver.spec.ts#L92-L94), where the beforeEach call clears all Subscriptions from the database. If this line is run while another test checks for the count or existence of subscriptions, it will fail. We only provided a solution to the existing problem by running the tests concurrently.

One alternative would be to connect to a different copy of the database for each jest worker thread, so they don't interfere. According to the jest issue tracker, [this is not easily doable](https://github.com/jestjs/jest/issues/8708).

Mocking the Prisma client would be possible, but we would lose quite a bit of testability there, because for example testing if records are returned in a specific order would no longer be possible (or depend entirely on the mock).

File: apps/api-example/src/app/slack-mail-provider.ts [hunk 13]
```diff
@@ -44,4 +47,19 @@ export class SlackMailProvider extends BaseMailProvider {
...
+      {
+        name: 'SlackEmptyTemplate',
+        uniqueIdentifier: 'slack-empty',
+        createdAt: new Date(),
+        updatedAt: new Date()
+      }
+    ]
+  }
+
+  getTemplateUrl(template: MailTemplate): string {
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply apps/api-example/src/app/slack-mail-provider.ts:62]: I prefer to leave it in the interface, since all other implementations need it. Only the Slack provider is a special case (it has no way to even define templates, and therefore no URLs either). Removing the method from the interface would mean that we have to check for its presence for all other providers (and remember implementing it if we ever add new providers).

File: docs/prisma.md [hunk 14]
```diff
@@ -181,6 +181,13 @@ You can reset the database by running the following command:
 npx prisma migrate reset
 ```
 
+## Prototype Prisma schema
+```sh
+npx prisma db push
+```
+
+Also see https://www.prisma.io/docs/concepts/components/prisma-migrate/db-push
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply docs/prisma.md:189]: When working on the schema, we found it was the easiest to modify the schema and push the schema if one wants to quickly test a change, because it removes the need to answer any interactive prompts in the console and doesn't create any temporary migration files.

File: libs/api/prisma/schema.prisma [hunk 15]
```diff
@@ -412,17 +414,22 @@ enum MailLogState {
...
+  id                      String   @id @default(cuid())
+  createdAt               DateTime @default(now())
+  modifiedAt              DateTime @updatedAt
+
+  recipientID             String
+  recipient               User   @relation(fields: [recipientID], references: [id], onDelete: Cascade)
+  state                   MailLogState
+  sentDate                DateTime
+  mailProviderID          String
+  mailIdentifier          String @unique
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/api/prisma/schema.prisma:426]: @tomaszdurka exactly.

File: libs/api/src/lib/graphql/article/article.private-mutation.ts [hunk 16]
```diff
@@ -165,6 +165,7 @@ export const duplicateArticle = async (
 
   const input: Prisma.ArticleRevisionCreateInput = {
     ...articleRevision,
+    blocks: articleRevision.blocks || Prisma.JsonNull,
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/api/src/lib/graphql/article/article.private-mutation.ts:168]: There were multiple TypeScript errors that prevented us from running the tests. We would have preferred not to meddle in those (unrelated) files, but it was required for running our tests.

Strangely, running the global test script does not  report the errors, but they definitely are valid TypeScript errors.

File: libs/api/tsconfig.spec.json [hunk 17]
```diff
@@ -1,6 +1,7 @@
 {
   "extends": "./tsconfig.json",
   "compilerOptions": {
+    "strict": false,
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/api/tsconfig.spec.json:4]: Generally, yes, this would be dangerous. We had to add it though, to keep the current behaviour. We did the following:

- Because the root tsconfig.base.json did not include the "strict" setting (which defaults to `false`), Prisma generated wrong type definitions for `findUnique`. FindUnique returns the record or null, but Prisma generated a signature that always returned the record.
- We added `"strict": true` to tsconfig.base.json: https://github.com/wepublish/wepublish/commit/bc0b48d712efba82054e4fca166391e4e4db582e#diff-0b280a445be167f54cd916c0718c952b8e0d4ca9621903d05eec25efd4c6ed6d
- We had to explicitely set `strict` to false in the `tsconfig.spec.json` (note: only for the specs), because the specs contained a huge amount of type errors.

So to summarize, the above change keeps the existing config for the `specs`, but allows us to enable strict checks for the rest of the code base.

File: libs/api/src/lib/graphql/mutation.public.ts [hunk 18]
```diff
@@ -704,18 +705,12 @@ export const GraphQLPublicMutation = new GraphQLObjectType<undefined, Context>({
...
-          type: SendMailType.LoginLink,
-          recipient: user.email,
-          data: {
-            url: urlAdapter.getLoginURL(token),
-            user
-          }
+          externalMailTemplateId: remoteTemplate,
+          recipient: user,
+          optionalData: {},
+          mailType: mailLogType.UserFlow
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/api/src/lib/graphql/mutation.public.ts:713]: @elias-summermatter This comes from [this commit](https://github.com/wepublish/wepublish/commit/e9829c96815d3828e276b7ffb3e8fa5a77f09451#diff-5b1c34d5c556f040853f4b86859e134a6e40380c3078eb4b6b9a79809966146b).
It looks like an error, can you confirm that?

File: libs/website/api/src/lib/graphql.ts [hunk 19]
```diff
@@ -1,123 +1,123 @@
...
-export type Exact<T extends {[key: string]: unknown}> = {[K in keyof T]: T[K]}
-export type MakeOptional<T, K extends keyof T> = Omit<T, K> & {[SubKey in K]?: Maybe<T[SubKey]>}
-export type MakeMaybe<T, K extends keyof T> = Omit<T, K> & {[SubKey in K]: Maybe<T[SubKey]>}
-const defaultOptions = {} as const
+import { gql } from '@apollo/client';
+import * as Apollo from '@apollo/client';
+export type Maybe<T> = T | null;
+export type InputMaybe<T> = Maybe<T>;
+export type Exact<T extends { [key: string]: unknown }> = { [K in keyof T]: T[K] };
+export type MakeOptional<T, K extends keyof T> = Omit<T, K> & { [SubKey in K]?: Maybe<T[SubKey]> };
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply libs/website/api/src/lib/graphql.ts:8]: No, we didn't modify it by hand, but I just checked: when I run the server, the file is generated with the spaces and semicolons. Maybe an eslint config?

File: tests/e2e.ts [hunk 20]
```diff
@@ -1,6 +1,6 @@
 import {ClientFunction, Role, Selector} from 'testcafe'
-import {slugify} from '../config/utilities'
 import * as process from 'process'
+import {slugify} from '../libs/utils/src'
```
[the remark this answers is not in the document]
Comment (LukasSkywalker) [reply tests/e2e.ts:3]: The @wepublish module is only available inside the `apps` and `libs` folder. Since this file is in `tests`, we must load `slugify` directly.

File: .gitignore [hunk 21]
```diff
@@ -8,6 +8,7 @@ npm-debug.log
 coverage/
 screenshots
 .env.local
+.env
```
Comment (Itrulia) [inline .gitignore:11]: .env is supposed to be commited, `.env.local` is the one, one should edit.
